data_science/archive_exploration/01 - subject coocurrence.ipynb

```diff
@@ -0,0 +1,320 @@
+    "\n",
+    "def cartesian(*arrays):\n",
+    "    return np.array([x.reshape(-1) for x in np.meshgrid(*arrays)]).T\n",
+    "\n",
+    "def clean_subject(subject):\n",
+    "    return subject.strip().lower().replace('<p>', '')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "let's load up our CALM data (exported in its entirity as a single `.json`, where each line is a record)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.read_json('data/calm_records.json')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "len(df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df.astype(str).describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Exploring individual columns\n",
+    "At the moment I have no idea what kind of information CALM contains - lets look at the list of column names"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "list(df)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here I'm looking through a sample of values in each column, choosing the columns to explore based on the their headings, a bit of contextual info from colleagues and the `df.describe()` above. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "df['Subject']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### After much trial and error...\n",
+    "Subjects look like an interesting avenue to explore further. Where subjects have _actually_ been filled in and the entry is not `None`, a list of subjects is returned.  \n",
+    "We can explore some of these subjects' subtleties by creating an adjacency matrix. We'll count the number of times each subject appears alongside every other subject and return a big $n \\times n$ matrix, where $n$ is the total number of unique subjects.  \n",
+    "We can use this adjacency matrix for all sorts of stuff, but we have to build it first. To start, lets get a uniqur list of all subjects. This involves unpacking each sub-list and flattening them out into one long list, before finding the unique elements. We'll then use the `clean_subject` function defined above to get rid of any irregularities which might become annoying later on."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "subjects = list(set(flatten(df['Subject'].dropna().tolist())))\n",
+    "clean_subjects = list(map(clean_subject, subjects))"
```

Comment:
Maybe I’m not following – if we want a list of unique subjects, shouldn’t we call `set()` only after we've normalised?

For example:

```pycon
>>> dirty_subjects = [['foo', 'bar'], ['FOO', 'Bar'], ['<p>Foo']]
>>> subjects = list(set(flatten(dirty_subjects)))
>>> list(map(clean_subject, subjects))
['foo', 'bar', 'bar', 'foo', 'foo']
```

vs

```pycon
>>> dirty_subjects = [['foo', 'bar'], ['FOO', 'Bar'], ['<p>Foo']]
>>> subjects = flatten(dirty_subjects)
>>> list(set(map(clean_subject, subjects)))
['foo', 'bar']
```